src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,511 @@ namespace nbl::asset
+                        const auto& specConstantFound = specConstants->find(count.specID);
+                        if (specConstantFound == specConstants->end())
+                            return false;
 
-    SPushConstantRange rngToPush; rngToPush.offset = 0u; rngToPush.size = 0u;
-    uint32_t stageFlags = 0u;
-    for (uint32_t i = 0u; i < ICPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE/sizeof(uint32_t); i += sizeof(uint32_t))
-    {
-        SPushConstantRange curr; curr.offset = i; curr.size = sizeof(uint32_t);
+                        descInfo.count = count.count;
+                    }
+                    else
+                    {
+                        descInfo.count = stageIntroBindingInfo.count.count;
+                    }
 
-        uint32_t tmpFlags = 0u;
-        for (uint32_t stg=0u; stg<infos.size(); stg++)
-            for (const SPushConstantRange& rng : ranges[stg])
-                if (curr.overlap(rng))
-                    tmpFlags |= rng.stageFlags;
+                    if (descInfo.count == 0u)
+                        return false;
 
-        if (!i)
-            stageFlags = tmpFlags;
+                    descInfo.isRuntimeSizedFlag = false;
+                }
+            }
+            else
+            {
+                descInfo.count = 1u;
+                descInfo.isRuntimeSizedFlag = false;
+            }
 
-        if (!tmpFlags)
-            continue;
-        if (tmpFlags == stageFlags)
-            rngToPush.size += sizeof(uint32_t);
-        else
-        {
-            rngToPush.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-            merged.push_back(rngToPush);
-            stageFlags = 0u;
+            const auto& pplnIntroDataFoundBinding = m_descriptorSetBindings[i].find(descInfo);
+            if (pplnIntroDataFoundBinding != m_descriptorSetBindings[i].end())
+            {
+                if (pplnIntroDataFoundBinding->type != stageIntroBindingInfo.type)
+                    return false;
+                descInfo.count = std::max(pplnIntroDataFoundBinding->count, descInfo.count);
+            }
 
-            rngToPush.offset = i;
-            rngToPush.size = sizeof(uint32_t);
+            highestBindingsTmp[i] = std::max<const int32_t>(highestBindingsTmp[i], stageIntroBindingInfo.binding);
+            descriptorsToMerge[i].insert(descInfo);
         }
     }
-    if (stageFlags)
+
+    // validate if only descriptors with the highest bindings are run-time sized
+    for (uint32_t i = 0u; i < ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
     {
-        rngToPush.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-        merged.push_back(rngToPush);
+        const auto& introBindingInfos = stageData->getDescriptorSetInfo(i);
+
+        for (const auto& descriptor : descriptorsToMerge[i])
+        {
+            if (descriptor.binding < highestBindingsTmp[i] && descriptor.isRuntimeSized())
+                return false;
```

Comment:
ok here you're checking if the new descriptors you're adding are not runtime sized if they're not the last.

But also, if you add new descriptors and `m_descriptorSetBindings` has an old runtime descriptor, you need to check whether it can still be the runtime descriptor

you can do this by looking up `m_highestBindingsNumbers[i]` (BEFORE you assign the new value) in `m_descriptorSetBindings`, checking if that descriptor was runtime sized and if it was you need to check if `highestBindingsTmp[i]` is higher